test/models/encoding.spec.ts

```diff
@@ -0,0 +1,161 @@
+import {
+  decodeOrders,
+  decodeIndexedOrders,
+  encodeOrders,
+  encodeIndexedOrders,
+} from "../../src";
+import BN from "bn.js";
+import { expect } from "chai";
+import "mocha";
+
+function hex(bytes: Uint8Array): string {
+  return `0x${Array.prototype.map
+    .call(bytes, (byte) => byte.toString(16).padStart(2, "0"))
+    .join("")}`;
+}
+
+function json(obj: unknown): unknown {
+  return JSON.parse(JSON.stringify(obj));
+}
+
+describe("Encoding Orders", () => {
+  describe("decodeOrders", () => {
+    it("accepts empty bytes and null and as input", () => {
+      expect(decodeOrders("")).to.deep.equal([]);
+      expect(decodeOrders("0x")).to.deep.equal([]);
+      expect(decodeOrders(null)).to.deep.equal([]);
+    });
+  });
+
+  describe("encodeOrders", () => {
+    it("encodes orders that can be decoded into the same values", () => {
+      const orders = [
+        {
+          user: "0x000102030405060708090a0b0c0d0e0f10111213",
+          sellTokenBalance: new BN(1),
+          buyToken: 2,
+          sellToken: 3,
+          validFrom: 4,
+          validUntil: 5,
+          priceNumerator: new BN("1000000000000000000"),
+          priceDenominator: new BN("1000000000000000000"),
+          remainingAmount: new BN(42),
+        },
+        {
+          user: "0x131211100f0e0d0c0b0a09080706050403020100",
+          sellTokenBalance: new BN(1),
+          buyToken: 2,
+          sellToken: 3,
+          validFrom: 4,
+          validUntil: 5,
+          priceNumerator: new BN(1337),
+          priceDenominator: new BN(0x12345678),
+          remainingAmount: new BN(42),
+        },
+      ];
+      const encoded = encodeOrders(orders);
+      const decoded = decodeOrders(hex(encoded));
+      expect(json(decoded)).to.deep.equal(json(orders));
```

Comment:
do we need to compare the json here or could we just deep compare the objects directly?

Reply:
Because different methods for constructing BN lead to different internal representations of the same number. Specifically, constructing from a number directly creates a BN with a single digit, while constructing from a hex string adds leading 0 digits:

```
> assert.deepEqual(new BN(0x2a), new BN("2a", 16))
Uncaught AssertionError [ERR_ASSERTION]: Expected values to be loosely deep-equal:

BN {
  length: 1,
  negative: 0,
  red: null,
  words: [
    42
  ]
}

should loosely deep-equal

BN {
  length: 1,
  negative: 0,
  red: null,
  words: [
    42,
    0
  ]
}
```

Since `BN` provides a `toJSON` implementation, it gets properly converted to a normalized implementation when comparing the JSON objects:
```
> assert.deepEqual(JSON.stringify(new BN(0x2a)), JSON.stringify(new BN("2a", 16)))
undefined
```

Reply:
This is why I'm all for dropping `BN` for `BigInt` ;)

Reply:
> This is why I'm all for dropping BN for BigInt ;)

+1 Maybe we can create a cleanup task for that?

Reply:
Thats gonna be a bigun'